### PR TITLE
Report AdditionalLocations for CS8618 (uninitialized non-nullable member)

### LINKED ISSUE #58073: AdditionalLocations for CS8618 in DiagnosticSuppressor
In EF Core, https://github.com/dotnet/efcore/issues/21608 introduced a DiagnosticsSuppressor to suppress CS8618 (non-nullable property is uninitialized) in a specific safe scenario. In classes without a constructor, the CS8618 diagnostic is reported on the property itself, so the suppressor can look at that property and function correctly. If a constructor is defined, however, the diagnostic is reported on the constructor declaration instead of the property, and the suppressor needs some way to resolve the uninitialized property.

[Diagnostic](https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.diagnostic?view=roslyn-dotnet-4.0.1) seems to have an [AdditionalLocations](https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.diagnostic.additionallocations?view=roslyn-dotnet-4.0.1#Microsoft_CodeAnalysis_Diagnostic_AdditionalLocations) property which looks right for this, but it's empty (because not populated for CS8618? Because it's a DiagnosticsSuppressor rather than a regular analyzer?).

Am I missing something? What would be the correct, safe way to figure out which non-nullable property is uninitialized from CS8618 in a DiagnosticsSuppressor?

Tracking issue on the EF Core side: https://github.com/dotnet/efcore/issues/26879

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #58073

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/UninitializedNonNullableFieldTests.cs
Comment (roji) [inline src/Compilers/CSharp/Test/Semantic/Semantics/UninitializedNonNullableFieldTests.cs:2502]: @Youssef1313 not sure the additional location makes sense when the diagnostic is issued on the property (i.e. no constructor), but I'm not sure what the conventions are here.

Comment (Youssef1313) [reply src/Compilers/CSharp/Test/Semantic/Semantics/UninitializedNonNullableFieldTests.cs:2502]: I'll let someone from the compiler team weigh on how this should behave. My intention was to allow the use of `AdditionalLocations` in all cases.

**Comment by Youssef1313** [comment]
Closing and re-opening for a new build

File: src/Compilers/CSharp/Test/Semantic/Semantics/UninitializedNonNullableFieldTests.cs
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/UninitializedNonNullableFieldTests.cs:2502]: I think not having to think about whether to use AdditionalLocations or not is the correct move.

**Review by 333fred — APPROVED** [review]
LGTM (commit 1). @dotnet/roslyn-compiler for a second review.

**Comment by Youssef1313** [comment]
@jcouv @RikkiGibson Can I get a second review here? Thanks!

File: src/Compilers/CSharp/Test/Semantic/Semantics/UninitializedNonNullableFieldTests.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/UninitializedNonNullableFieldTests.cs:2477]: It's slightly unfortunate that this generated baseline doesn't include the additional locations.

**Review by RikkiGibson — APPROVED** [review]

**Comment by RikkiGibson** [comment]
Thanks for the contribution @Youssef1313!

**Comment by roji** [comment]
@RikkiGibson and others, I'm testing with dotnet SDK 7.0.100-preview.3.22179.4 and the AdditionalLocations property on the diagnostic is empty - can you confirm this came out with perview3?

(it's always a bit difficult to understand which SDK version the VS version maps to)

**Comment by RikkiGibson** [comment]
This change should definitely be in preview3. What's the scenario/environment where you observe that AdditionalLocations is empty?

**Comment by roji** [comment]
Apologies, I can no longer reproduce the issue and am seeing the populated AdditionalLocations (possibly a caching issue after upgrading or something). Thanks for the quick response.

**Comment by roji** [comment]
@RikkiGibson sorry, one last question... If I'm understanding everything correctly, I can see AdditionalLocations populated when referencing Microsoft.CodeAnalysis.CSharp [4.2.0-2.final](https://www.nuget.org/packages/Microsoft.CodeAnalysis.CSharp/4.2.0-2.final), but not when referencing [4.1.0](https://www.nuget.org/packages/Microsoft.CodeAnalysis.CSharp/4.1.0). Is 4.2.0-2.final the equivalent of the Roslyn that's included in SDK 7.0.0-preview.3, and will 4.2.0 be released at the same time as .NET 7.0?

**Comment by RikkiGibson** [comment]
> Is 4.2.0-2.final the equivalent of the Roslyn that's included in SDK 7.0.0-preview.3

It looks like preview3 has some build of [4.2.0-4](https://github.com/dotnet/sdk/pull/24774). The way I answer these questions is by querying [PRs to dotnet/sdk](https://github.com/dotnet/sdk/pulls?q=is%3Apr+is%3Amerged+roslyn+base%3Arelease%2F7.0.1xx-preview3).

> will 4.2.0 be released at the same time as .NET 7.0?

4.2.0 will be released sooner than .NET 7.0, I think. See the (internal) [dev17.2 schedule](https://dev.azure.com/devdiv/DevDiv/_wiki/wikis/DevDiv.wiki/24116/Dev17.2-LTS-Schedule-Details) and [NET 7 schedule](https://dev.azure.com/devdiv/DevDiv/_wiki/wikis/DevDiv.wiki/26302/NET-7-Schedule).

**Comment by roji** [comment]
Thanks again @RikkiGibson.

**Pull request merged**
